packages/core/src/extensions/Blocks/api/customBlocks.ts

```diff
@@ -0,0 +1,164 @@
+import {
+  createInternalBlockSpec,
+  createStronglyTypedTiptapNode,
+  getBlockFromPos,
+  parse,
+  propsToAttributes,
+  wrapInBlockStructure,
+} from "./block";
+import {
+  BlockFromBlockConfig,
+  BlockSchemaWithBlock,
+  PropSchema,
+} from "./blockTypes";
+
+// Defines the config for a custom block. Meant to be used as an argument to
+// `createBlockSpec`, which will create a new block spec from it.
+export type CustomBlockConfig = {
+  type: string;
+  readonly propSchema: PropSchema;
+  containsInlineContent: boolean;
+};
+
+export type BlockConfigFromCustomBlockConfig<T extends CustomBlockConfig> =
+  Omit<T, "containsInlineContent"> & {
+    content: T["containsInlineContent"] extends true ? "inline" : "none";
+  };
+
+export type BlockFromCustomBlockConfig<T extends CustomBlockConfig> =
+  BlockFromBlockConfig<BlockConfigFromCustomBlockConfig<T>>;
+
+export type CustomBlockImplementation<T extends CustomBlockConfig> = {
```

Comment:
CustomBlockImplementation eventually maps to a BlockImplementation indeed. 

Renamed to `BlockImplementation` to `TiptapBlockImplementation` because it's the type that eventually has a tiptap node